### PR TITLE
 Adding support to construct pinot java client with pinot controller url

### PR DESCRIPTION
**Pull request by xiangfu0**
## Description
Adding support to construct pinot java client with pinot controller url.

## Release Notes
Adding support to construct pinot java client with pinot controller url.

### CODE REVIEW

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBasedBrokerSelector.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            controllerUrls[controllerIdx], this.controllerFetchRetriesIntervalMills), e);
+        controllerIdx = (controllerIdx + 1) % controllerUrls.length;
+        try {
+          Thread.sleep(controllerFetchRetriesIntervalMills);
+        } catch (InterruptedException interruptedException) {
+          // Swallow
+        }
+      }
+    }
+    LOGGER.warn(
```
Comment (mayankshriv) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBasedBrokerSelector.java:118]: Same here, `String.format` should be avoidable.

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBasedBrokerSelector.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        String.format("Failed to fetch controller broker information with %d retries", this.controllerFetchRetries));
+    return null;
+  }
+
+  @Override
+  public String selectBroker(String table) {
+    if (table == null) {
+      List<String> list = allBrokerListRef.get();
```
Comment (mayankshriv) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBasedBrokerSelector.java:126]: `allBrokerListRef` being computed outside of this method makes the api vulnerable to getting a broker that is no longer ONLINE in EV.

Comment (fx19880617) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBasedBrokerSelector.java:126]: This is true, I'm caching the broker list with a scheduled call.  I don't want to let each query calling controller very time.

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBasedBrokerSelector.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.commons.io.IOUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.pinot.client.ExternalViewReader.OFFLINE_SUFFIX;
```
[the remark this answers is not in the document]
Comment (fx19880617) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBasedBrokerSelector.java]: done.

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBasedBrokerSelector.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+  private static final Logger LOGGER = LoggerFactory.getLogger(ControllerBasedBrokerSelector.class);
+
+  public static final int DEFAULT_CONTROLLER_REQUEST_RETRIES = 3;
+  public static final long DEFAULT_CONTROLLER_REQUEST_RETRIES_INTERVAL_IN_MILLS = 5000; // 5 seconds
+  public static final long DEFAULT_CONTROLLER_REQUEST_SCHEDULE_INTERVAL_IN_MILLS = 300000; // 5 minutes
+
+  private static final String CONTROLLER_ONLINE_TABLE_BROKERS_MAP_URL_TEMPLATE = "%s/brokers/tables?state=online";
+  private static final Random RANDOM = new Random();
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+  private final String[] controllerUrls;
```
[the remark this answers is not in the document]
Comment (fx19880617) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBasedBrokerSelector.java]: done.

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBasedBrokerSelector.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    int controllerIdx = RANDOM.nextInt(controllerUrls.length);
+    for (int i = 0; i < controllerFetchRetries; i++) {
+      try {
+        String brokerTableQueryResp = IOUtils.toString(
+            new URL(String.format(CONTROLLER_ONLINE_TABLE_BROKERS_MAP_URL_TEMPLATE, controllerUrls[controllerIdx])));
+        Map<String, List<String>> tablesToBrokersMap = OBJECT_MAPPER.readValue(brokerTableQueryResp, Map.class);
+        for (String table : tablesToBrokersMap.keySet()) {
+          List<String> brokerHostPorts = new ArrayList<>();
+          for (String broker : tablesToBrokersMap.get(table)) {
+            brokerHostPorts.add(broker.replace("Broker_", "").replace("_", ":"));
```
[the remark this answers is not in the document]
Comment (fx19880617) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBasedBrokerSelector.java]: Moved to a util function. This is also used in other functions.

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBasedBrokerSelector.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+          }
+          tablesToBrokersMap.put(table, brokerHostPorts);
+        }
+        return tablesToBrokersMap;
+
+      } catch (Exception e) {
+        LOGGER.warn(String.format("Unable to fetch controller broker information from '%s', retry in %d millseconds",
```
[the remark this answers is not in the document]
Comment (fx19880617) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBasedBrokerSelector.java]: Updated.

**Comment by kishoreg** [comment]
@mayankshriv @Jackie-Jiang please review this?

**Comment by kishoreg** [comment]
@fx19880617 is this still valid or @KKcorps has already incorporated this?

**Pull request closed without merging**